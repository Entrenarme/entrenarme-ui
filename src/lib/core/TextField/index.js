// @flow
import * as React from 'react';
import MTextField from '@material-ui/core/TextField';
import InputAdornment from '@material-ui/core/InputAdornment';
import {
  createMuiTheme,
  MuiThemeProvider,
  createGenerateClassName,
  jssPreset,
} from '@material-ui/core/styles';
import JssProvider from 'react-jss/lib/JssProvider';
import { create } from 'jss';
import styled from 'styled-components';
import debounce from 'lodash/debounce';

import colors from '../../helpers/colors';
import { regularFont } from '../../helpers/fonts';

type Props = {
  label?: string,
  placeholder?: React.Node,
  adornment?: React.Node,
  type?: 'text' | 'number' | 'password' | 'email',
  onChange?: Function,
  debounceMs?: number,
};

const theme = createMuiTheme({
  palette: {
    primary: { main: colors.secondary.default, dark: colors.secondary.default },
    error: { main: colors.error.default, dark: colors.error.default },
    text: { primary: colors.gray.placeholder },
  },
});

const generateClassName = createGenerateClassName({ productionPrefix: 'eui' });
const jss = create(jssPreset());

const STextField = styled(MTextField)`
  > div:before {
    border-bottom: 1px solid ${colors.gray.placeholder};
  }

  input {
    color: ${colors.text} !important;
    font-family: ${regularFont} !important;
  }
`;

class TextField extends React.Component<Props> {
  static defaultProps = {
    label: '',
    placeholder: '',
    adornment: null,
    type: 'text',
    onChange: () => {},
    debounceMs: 0,
  };

  onInputChange = debounce((e: SyntheticEvent<*>) => {
    this.props.onChange(e);
  }, this.props.debounceMs);

  onChange = e => {
    e.persist();
    this.onInputChange(e);
  };

  render() {
    const { adornment, onChange, debounceMs, ...rest } = this.props;
    return (
      <JssProvider jss={jss} generateClassName={generateClassName}>
        <MuiThemeProvider theme={theme}>
          <STextField
            onChange={this.onChange}
            {...rest}
            InputProps={{
              startAdornment: adornment ? (
                <InputAdornment position="start">{adornment}</InputAdornment>
              ) : null,
            }}
          />
        </MuiThemeProvider>
      </JssProvider>
    );
  }
}

export default TextField;
