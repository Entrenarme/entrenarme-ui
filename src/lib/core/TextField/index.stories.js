// @flow
import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSearch } from '@fortawesome/free-solid-svg-icons';

import TextField from './index';
import colors from '../../helpers/colors';

const actions = {
  onChange: action('onChange'),
};

storiesOf('TextField', module)
  .add('default', () => <TextField {...actions} />)
  .add('with placeholder', () => (
    <TextField {...actions} placeholder="Search" />
  ))
  .add('with adornment at the start', () => (
    <TextField
      {...actions}
      adornment={
        <FontAwesomeIcon icon={faSearch} color={colors.gray.placeholder} />
      }
    />
  ))
  .add('with debounce', () => <TextField {...actions} debounceMs={400} />)
  .add('with all together', () => (
    <TextField
      {...actions}
      placeholder="Search"
      debounceMs={400}
      adornment={
        <FontAwesomeIcon icon={faSearch} color={colors.gray.placeholder} />
      }
    />
  ));
