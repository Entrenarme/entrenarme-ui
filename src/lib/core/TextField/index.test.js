import * as React from 'react';
import { render, fireEvent, wait } from 'react-testing-library';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSearch } from '@fortawesome/pro-light-svg-icons';

import TextField from './index';
import { debounce } from '../../../../node_modules/rxjs/operator/debounce';

test('renders default', () => {
  const { container } = render(<TextField />);
  expect(container.firstChild).toMatchSnapshot();
});

test('renders with placeholder', () => {
  const { container } = render(<TextField placeholder="placeholder" />);
  expect(container.firstChild).toMatchSnapshot();
});

test('renders with label', () => {
  const { container } = render(<TextField label="label" />);
  expect(container.firstChild).toMatchSnapshot();
});

test('renders with adornment start', () => {
  const { container } = render(
    <TextField adornment={<FontAwesomeIcon icon={faSearch} />} />,
  );
  expect(container.firstChild).toMatchSnapshot();
});

test('onChange gets the event', async () => {
  const onChange = jest.fn();
  const newValue = 'wifi';
  const { getByPlaceholderText } = render(
    <TextField placeholder="placeholder" onChange={onChange} />,
  );
  const textField = getByPlaceholderText('placeholder');
  textField.value = newValue;
  fireEvent.change(textField);

  const begin = Date.now();
  await wait(() => {
    return expect(onChange).toHaveBeenCalledTimes(1);
  });
  const end = Date.now();
  expect(end - begin).toBeLessThanOrEqual(50);
  expect(textField.value).toBe(newValue);
});

test('onChange gets the event debounced', async () => {
  const onChange = jest.fn();
  const newValue = 'wifi';
  const debounceMs = 300;
  const { getByPlaceholderText } = render(
    <TextField
      placeholder="placeholder"
      debounceMs={debounceMs}
      onChange={onChange}
    />,
  );
  const textField = getByPlaceholderText('placeholder');
  textField.value = newValue;
  fireEvent.change(textField);

  const begin = Date.now();
  await wait(() => {
    return expect(onChange).toHaveBeenCalledTimes(1);
  });
  const end = Date.now();
  expect(end - begin).toBeGreaterThanOrEqual(debounceMs);
  expect(textField.value).toBe(newValue);
});
