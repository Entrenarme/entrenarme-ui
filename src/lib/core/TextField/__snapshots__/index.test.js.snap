// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default 1`] = `
<div
  class="MuiFormControl-root-1 sc-bdVaJa faWPZi"
>
  <div
    class="MuiInput-root-5 MuiInput-formControl-6 MuiInput-underline-9"
  >
    <input
      aria-invalid="false"
      class="MuiInput-input-13"
      placeholder=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`renders with adornment start 1`] = `
<div
  class="MuiFormControl-root-62 sc-bdVaJa faWPZi"
>
  <div
    class="MuiInput-root-66 MuiInput-formControl-67 MuiInput-underline-70"
  >
    <div
      class="MuiInputAdornment-root-79 MuiInputAdornment-positionStart-80"
    >
      <svg
        aria-hidden="true"
        data-prefix="fas"
        data-icon="search"
        class="svg-inline--fa fa-search fa-w-16 "
        role="img"
        xmlns="http://www.w3.org/2000/svg"
        viewBox="0 0 512 512"
      >
        <path
          fill="currentColor"
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        />
      </svg>
    </div>
    <input
      aria-invalid="false"
      class="MuiInput-input-74"
      placeholder=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`renders with label 1`] = `
<div
  class="MuiFormControl-root-35 sc-bdVaJa faWPZi"
>
  <label
    class="MuiFormLabel-root-44 MuiInputLabel-root-39 MuiInputLabel-formControl-40 MuiInputLabel-animated-43"
    data-shrink="false"
  >
    label
  </label>
  <div
    class="MuiInput-root-49 MuiInput-formControl-50 MuiInput-underline-53"
  >
    <input
      aria-invalid="false"
      class="MuiInput-input-57"
      placeholder=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`renders with placeholder 1`] = `
<div
  class="MuiFormControl-root-18 sc-bdVaJa faWPZi"
>
  <div
    class="MuiInput-root-22 MuiInput-formControl-23 MuiInput-underline-26"
  >
    <input
      aria-invalid="false"
      class="MuiInput-input-30"
      placeholder="placeholder"
      type="text"
      value=""
    />
  </div>
</div>
`;
